File: app/src/main/java/ca/judacribz/gainzassist/activities/how_to_videos/ThumbnailAdapter.java
package ca.judacribz.gainzassist.activities.how_to_videos;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import java.util.ArrayList;

import ca.judacribz.gainzassist.R;

public class ThumbnailAdapter extends RecyclerView.Adapter<ThumbnailViewHolder> implements
        ThumbnailViewHolder.VideoClickObserver {

    // Interfaces
    // --------------------------------------------------------------------------------------------
    private VideoClickObserver videoClickObserver;

    public interface VideoClickObserver {
        void onVideoClick(String videoId);
    }

    void setVideoClickObserver(VideoClickObserver videoClickObserver) {
        this.videoClickObserver = videoClickObserver;
    }

    // Global Vars
    // --------------------------------------------------------------------------------------------
    private ArrayList<String> videoIds;
    private ArrayList<String> videoTitles;
    // --------------------------------------------------------------------------------------------

    // ######################################################################################### //
    // ThumbnailAdapter Constructor                                                              //
    // ######################################################################################### //
    ThumbnailAdapter(ArrayList<String> videoIds, ArrayList<String> videoTitles) {
        this.videoIds = videoIds;
        this.videoTitles = videoTitles;
    }
    // ######################################################################################### //


    // RecyclerView.Adapter<ThumbnailViewHolder> Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @NonNull
    @Override
    public ThumbnailViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.part_thumbnail, parent, false);

        ThumbnailViewHolder thumbnailViewHolder = new ThumbnailViewHolder(view, videoIds);
        thumbnailViewHolder.setVideoClickObserver(this);

        return thumbnailViewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ThumbnailViewHolder holder, int position) {
        holder.bind(videoIds.get(position), videoTitles.get(position));
    }

    @Override
    public int getItemCount() {
        return videoIds.size();
    }
    //RecyclerView.Adapter<ThumbnailViewHolder>//Override//////////////////////////////////////////


    // Interface Callback: ThumbnailViewHolder.VideoClickObserver Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void onVideoClick(String videoId) {
        videoClickObserver.onVideoClick(videoId);
    }
    //ThumbnailViewHolder.VideoClickObserver//Override/////////////////////////////////////////////
}

File: app/src/main/java/ca/judacribz/gainzassist/activities/how_to_videos/ThumbnailAdapter.kt
package ca.judacribz.gainzassist.activities.how_to_videos

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import ca.judacribz.gainzassist.R

// --------------------------------------------------------------------------------------------
// ######################################################################################### //
// ThumbnailAdapter Constructor                                                              //
// ######################################################################################### //
class ThumbnailAdapter internal constructor(
    private val videoIds: ArrayList<String>,
    private val videoTitles: ArrayList<String>
) : RecyclerView.Adapter<ThumbnailViewHolder>(), ThumbnailViewHolder.VideoClickObserver {
    // Interfaces
    // --------------------------------------------------------------------------------------------
    private var videoClickObserver: VideoClickObserver? = null

    interface VideoClickObserver {
        fun onVideoClick(videoId: String?)
    }

    fun setVideoClickObserver(videoClickObserver: VideoClickObserver?) {
        this.videoClickObserver = videoClickObserver
    }

    // ######################################################################################### //
    // RecyclerView.Adapter<ThumbnailViewHolder> Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ThumbnailViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.part_thumbnail, parent, false)
        val thumbnailViewHolder = ThumbnailViewHolder(view, videoIds)
        thumbnailViewHolder.setVideoClickObserver(this)
        return thumbnailViewHolder
    }

    override fun onBindViewHolder(holder: ThumbnailViewHolder, position: Int) {
        holder.bind(videoIds[position], videoTitles[position])
    }

    override fun getItemCount(): Int {
        return videoIds.size
    }

    //RecyclerView.Adapter<ThumbnailViewHolder>//Override//////////////////////////////////////////
    // Interface Callback: ThumbnailViewHolder.VideoClickObserver Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    override fun onVideoClick(videoId: String) {
        videoClickObserver!!.onVideoClick(videoId)
    } //ThumbnailViewHolder.VideoClickObserver//Override/////////////////////////////////////////////
}